core/src/main/java/com/linecorp/armeria/client/retry/RetryConfigBuilder.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.time.Duration;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.MoreObjects.ToStringHelper;
+
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.Response;
+
+/**
+ * Builds a {@link RetryConfig}.
+ * A {@link RetryConfig} instance encapsulates the used {@link RetryRule}, maxTotalAttempts,
+ * and responseTimeoutMillisForEachAttempt.
+ */
+public final class RetryConfigBuilder<T extends Response> {
+    int maxTotalAttempts = Flags.defaultMaxTotalAttempts();
+    long responseTimeoutMillisForEachAttempt = Flags.defaultResponseTimeoutMillis();
+    @Nullable final RetryRule retryRule;
+    @Nullable final RetryRuleWithContent<T> retryRuleWithContent;
+    final int maxContentLength;
+
+    private RetryConfigBuilder() {
+        throw new IllegalStateException("RetryConfigBuilder must have a rule.");
+    }
+
+    /**
+     * Returns a {@link RetryConfigBuilder} with this {@link RetryRule}.
+     */
+    public RetryConfigBuilder(RetryRule retryRule) {
+        this.retryRule = checkNotNull(retryRule);
```

Comment:
Global comments: Please use `requireNonNull` where `checkNotNull` is used.
```suggestion
        this.retryRule = requireNonNull(retryRule, "retryRule");
```